shell/platform/android/io/flutter/embedding/FlutterRenderer.java

```diff
@@ -0,0 +1,284 @@
+        surfaceTexture
+    );
+    registerTexture(entry.id(), surfaceTexture);
+    return entry;
+  }
+
+  final class SurfaceTextureRegistryEntry implements TextureRegistry.SurfaceTextureEntry {
+    private final long id;
+    private final SurfaceTexture surfaceTexture;
+    private boolean released;
+
+    SurfaceTextureRegistryEntry(long id, SurfaceTexture surfaceTexture) {
+      this.id = id;
+      this.surfaceTexture = surfaceTexture;
+      this.surfaceTexture.setOnFrameAvailableListener(new SurfaceTexture.OnFrameAvailableListener() {
+        @Override
+        public void onFrameAvailable(SurfaceTexture texture) {
+          markTextureFrameAvailable(SurfaceTextureRegistryEntry.this.id);
+        }
+      });
+    }
+
+    @Override
+    public SurfaceTexture surfaceTexture() {
+      return surfaceTexture;
+    }
+
+    @Override
+    public long id() {
+      return id;
+    }
+
+    @Override
+    public void release() {
+      if (released) {
+        return;
+      }
+      unregisterTexture(id);
+      surfaceTexture.release();
+      released = true;
+    }
+  }
+  //------ END TextureRegistry IMPLEMENTATION ----
+
+  // TODO(mattcarroll): what exactly is this method intended to do?
+  public void surfaceCreated(Surface surface, int backgroundColor) {
+    flutterJNI.nativeSurfaceCreated(nativeObjectReference, surface, backgroundColor);
+  }
+
+  // TODO(mattcarroll): what exactly is this method intended to do?
+  public void surfaceChanged(int width, int height) {
+    flutterJNI.nativeSurfaceChanged(nativeObjectReference, width, height);
+  }
+
+  // TODO(mattcarroll): what exactly is this method intended to do?
+  public void surfaceDestroyed() {
+    flutterJNI.nativeSurfaceDestroyed(nativeObjectReference);
+  }
+
+  // TODO(mattcarroll): what exactly is this method intended to do?
+  public void setViewportMetrics(float devicePixelRatio,
+                                 int physicalWidth,
+                                 int physicalHeight,
+                                 int physicalPaddingTop,
+                                 int physicalPaddingRight,
+                                 int physicalPaddingBottom,
+                                 int physicalPaddingLeft,
+                                 int physicalViewInsetTop,
+                                 int physicalViewInsetRight,
+                                 int physicalViewInsetBottom,
+                                 int physicalViewInsetLeft) {
+    flutterJNI.nativeSetViewportMetrics(
+        nativeObjectReference,
+        devicePixelRatio,
+        physicalWidth,
+        physicalHeight,
+        physicalPaddingTop,
+        physicalPaddingRight,
+        physicalPaddingBottom,
+        physicalPaddingLeft,
+        physicalViewInsetTop,
+        physicalViewInsetRight,
+        physicalViewInsetBottom,
+        physicalViewInsetLeft
+    );
+  }
+
+  // TODO(mattcarroll): why does this method exist?
+  public Bitmap getBitmap() {
```

Comment:
Are you doing this abstraction to make this class more testable by passing in a different implementation of FlutterJNI?

which brings me to the next point. It would be great to have tests for these to make sure they are doing the "right thing". :-).

Reply:
One goal is testability.  Another goal is facilitating a mental model of the Java implementation that doesn't require thinking about mixtures of Java and native calls.  I think it's easier to explain relationships when you don't have somewhat random jumps from Java code to C code.